lib/OIDCLogoutHandler.php

```diff
@@ -0,0 +1,77 @@
+<?php
+/**
+ * Created by PhpStorm.
+ * User: patrick
+ * Date: 12/21/17
+ * Time: 3:26 PM
+ */
+
+namespace SimpleSAML\Module\authoauth2;
+
+use League\OAuth2\Client\Provider\Exception\IdentityProviderException;
+use SimpleSAML\Logger;
+use SimpleSAML\Module;
+use SimpleSAML\Module\authoauth2\Auth\Source\OAuth2;
+use SimpleSAML\Module\authoauth2\Auth\Source\OpenIDConnect;
+use SimpleSAML\Utils\HTTP;
+
+class OIDCLogoutHandler
+{
+    private $expectedStageState = OpenIDConnect::STAGE_LOGOUT;
+    private $expectedStateAuthId = OAuth2::AUTHID;
+
+    private $expectedPrefix = OAuth2::STATE_PREFIX . '-';
+
+    /**
+     * Look at the state parameter returned by the OAuth2 server and determine if we can handle it;
+     * @return bool true if response can be handled by this module
+     */
+    public function canHandleResponse()
+    {
+        return $this->canHandleResponseFromRequest($_REQUEST);
+    }
+
+    public function canHandleResponseFromRequest(array $request)
+    {
+        return strpos(@$request['state'], $this->expectedPrefix) === 0;
+    }
+
+    /**
+     * Handle an OAuth2 response.
+     */
+    public function handleResponse()
+    {
+        $this->handleResponseFromRequest($_REQUEST);
+    }
+
+    public function handleResponseFromRequest(array $request)
+    {
+        Logger::debug('authoauth2: logout request=' . var_export($request, true));
+
+        if (!$this->canHandleResponseFromRequest($request)) {
+            // phpcs:ignore Generic.Files.LineLength.TooLong
+            throw new \SimpleSAML\Error\BadRequest('Either missing state parameter on OpenID Connect logout callback, or cannot be handled by authoauth2');
+        }
+        $stateIdWithPrefix = $request['state'];
+        $stateId = substr($stateIdWithPrefix, strlen($this->expectedPrefix));
+        //TODO: decide how no-state errors should be handled?
+        // Likely cause is user clicked back button (state was already consumed and removed) or session expired
+        $state = \SimpleSAML\Auth\State::loadState($stateId, $this->expectedStageState);
+
+        // Find authentication source
+        if (!array_key_exists($this->expectedStateAuthId, $state)) {
+            throw new \SimpleSAML\Error\BadRequest('No authsource id data in state for ' . $this->expectedStateAuthId);
+        }
+        $sourceId = $state[$this->expectedStateAuthId];
+
+        /**
+         * @var OAuth2 $source
+         */
+        $source = \SimpleSAML\Auth\Source::getById($sourceId, OpenIDConnect::class);
```

Comment:
What's `$source` used for? I don't see anything in the following lines using it, other than checking it exists. If that's the only thing we need, we might just skip the variable and use the call to `getById()` as the expression inside the `if`.

Reply:
The variable certainly isn't used, so it could be inlined in the if statement if wanted, but shouldn't it be used for the completeLogout call? I really don't understand how the state handling here is supposed to work...